homeassistant/components/esphome/dashboard.py

```diff
@@ -4,69 +4,153 @@
+
 
-    dashboard = ESPHomeDashboard(hass, addon_slug, url, async_get_clientsession(hass))
-    try:
+@callback
+def async_get_dashboard_manager(hass: HomeAssistant) -> ESPHomeDashboardManager | None:
+    """Get the dashboard manager if it already exists."""
+    manager: ESPHomeDashboardManager | None = hass.data.get(KEY_DASHBOARD_MANAGER)
+    return manager
+
+
+async def async_get_or_create_dashboard_manager(
+    hass: HomeAssistant,
+) -> ESPHomeDashboardManager:
+    """Get the dashboard manager or create it."""
+    if KEY_DASHBOARD_MANAGER not in hass.data:
+        manager = ESPHomeDashboardManager(hass)
+        await manager.async_setup()
+        hass.data[KEY_DASHBOARD_MANAGER] = manager
+    else:
+        manager = hass.data[KEY_DASHBOARD_MANAGER]
+    return manager
+
+
+class ESPHomeDashboardManager:
+    """Class to manage the dashboard and restore it from storage."""
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize the dashboard manager."""
+        self._hass = hass
+        self._store: Store[dict[str, Any]] = Store(hass, STORAGE_VERSION, STORAGE_KEY)
+        self._data: dict[str, Any] | None = None
+        self._current_dashboard: ESPHomeDashboard | None = None
+        self._cancel_shutdown: CALLBACK_TYPE | None = None
+
+    async def async_setup(self) -> None:
+        """Restore the dashboard from storage."""
+        self._data = await self._store.async_load()
+        if (data := self._data) and (info := data.get("info")):
+            await self.async_set_dashboard_info(
+                info["addon_slug"], info["host"], info["port"]
+            )
+
+    @callback
+    def async_get(self) -> ESPHomeDashboard | None:
+        """Get the current dashboard."""
+        return self._current_dashboard
+
+    async def async_set_dashboard_info(
+        self, addon_slug: str, host: str, port: int
+    ) -> None:
+        """Set the dashboard info."""
+        url = f"http://{host}:{port}"
+        hass = self._hass
+
+        if cur_dashboard := self._current_dashboard:
+            if cur_dashboard.addon_slug == addon_slug and cur_dashboard.url == url:
+                # Do nothing if we already have this data.
+                return
+            # Clear and make way for new dashboard
+            await cur_dashboard.async_shutdown()
+            if self._cancel_shutdown is not None:
+                self._cancel_shutdown()
+                self._cancel_shutdown = None
+            self._current_dashboard = None
+
+        dashboard = ESPHomeDashboard(
+            hass, addon_slug, url, async_get_clientsession(hass)
+        )
         await dashboard.async_request_refresh()
-    except UpdateFailed as err:
-        logging.getLogger(__name__).error("Ignoring dashboard info: %s", err)
-        return
+        if not dashboard.last_update_success:
+            _LOGGER.error("Ignoring dashboard info: %s", dashboard.last_exception)
+            return
```

Comment:
In this case we could have just cleared the old information, so you would still need to do a reload.

Reply:
changed the behavior to still do a reload if there was a previous dashboard df2e67f3754b3bbf729b890c015fbad6c4f2343d